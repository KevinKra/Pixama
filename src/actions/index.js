export const loginUser = (user) => ({
  type: "LOGIN_USER",
  user
});

export const logoutUser = () => ({
  type: "LOGOUT_USER"
});

export const addNewUser = (name, email, password) => ({
  type: "ADD_NEW_USER",
  name,
  email,
  password
});

export const getFavorites = favorites => ({
  type: "GET_FAVORITES",
  favorites
});

export const cleanPopularMovies = popMovies => ({
  type: "CLEAN_POPULAR_MOVIES",
  popMovies
});



