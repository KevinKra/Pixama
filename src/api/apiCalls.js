import API_KEY from "./apikey";

export const fetchPopularMovies = async () => {
  let response = await fetch(
    `https://api.themoviedb.org/3/discover/movie?certification_country=US&api_key=${API_KEY}`
  );
  const movies = await response.json();
  return movies.results;
};

//This method is not hooked up anywhere at this point
//data must be an object with keys of name, email and password
export const fetchNewUser = (url, data)  => {
  return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Email already exists", response.message)
    } else {
      return response.json()
    }
  })
}

//this method is being imported and called in LoginCard component
export const fetchUser = (url, data)  => {

  return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Not a valid login",response.message)
    } else {
      return response.json()
    }
  })
}

//Method to add a movie to a user's favorites, this method not hooked up to a component yet
// url in component method needs to be 'http://localhost:3000/api/users/favorites/new'
//data needs to be object like this: {movie_id: 550, user_id: 1, title: "Fight Club", poster_path: "/adw6Lq9FiC9zjYEpOqfq03ituwp.jpg" , release_date: "1999-10-15", vote_average:8.4, overview: "A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground \"fight clubs\" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion."}
//response will be the new favorite ID
export const postFavorite = (url, data)  => {
console.log(data)
return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Add error",response.message)
    } else {
      return response.json()
    }
  })
};


export const fetchFavorites = (url)  => {

return fetch(url)
  .then(response=> {
    if(!response.ok) {
      throw Error ("Couldn't get favorites",response.message)
    } else {
      return response.json()
    }
  })
};