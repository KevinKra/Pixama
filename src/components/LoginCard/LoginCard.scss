.login-card {
  min-width: 300px;
  min-height: 200px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  background-color: white;
  border: 1px solid red;

  input {
    width: 250px;
    height: 40px;
  }

  button {
    background-color: red;
  }
  
}

