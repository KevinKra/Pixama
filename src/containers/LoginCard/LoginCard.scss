.login-card {
  min-width: 300px;
  min-height: 200px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  background-color: black;
  position: fixed;
  bottom: 40vh;
  top: 20vh;
  left: 15vw;
  z-index: 5;
  width: 70%;
  height: 50vh;
  opacity: .85;
  padding: 60px 0px;
  box-shadow: 20px 20px 23px -8px rgba(0,0,0,0.75);


  input {
    width: 250px;
    height: 40px;
    border-radius: 4px;
    border: red solid 2px;
  }

  button {
    color: white;
    height: 45px;
    width: 150px;
    background-color: red;
    border-radius: 4px;
    border: red solid 2px;
    &:hover {
      cursor: pointer;
      transform: scale(1.05);
    }
  }

  .register-text {
    color: white;
    &:visited {
      color: white;
    }
  }

  .login-error-text {
    color: red;
  }
  
}

