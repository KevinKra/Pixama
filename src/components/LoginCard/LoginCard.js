import React, { Component } from "react";
import './LoginCard.scss'
import { fetchUser } from "../../api/apiCalls";


 class LoginCard extends Component {
  state = {
    email: '',
    password: '',
    userInfo: {},
    error: ''
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name] : e.target.value
    })
  }


onSubmit = async() => {
let userData = {email: this.state.email, password: this.state.password}

try {
  const userInfo = await fetchUser('http://localhost:3000/api/users', userData)
  this.setState(userInfo)
  console.log(this.state.data)
}catch (error) {
 this.setState({ error })
}

this.clearForm()
}


clearForm = () => {
  this.setState({
    email: '',
    password: ''
  })
}

  render() {
    return (
      <form className="login-card">
        <input onChange={this.handleChange} name="email" value={this.state.email} type="text" placeholder="Email"/>
        <input onChange={this.handleChange}name="password" value ={this.state.password} type="text" placeholder="Password"/>
        <button type="button" onClick={this.onSubmit}>Login</button>
      </form>
    );
  }

}




export default LoginCard;
