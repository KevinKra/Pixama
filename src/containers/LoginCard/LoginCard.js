import React, { Component } from "react";
import "./LoginCard.scss";
import { fetchUser, fetchFavorites } from "../../api/apiCalls";
import { connect } from "react-redux";
import { loginUser, getFavorites } from "../../actions";
import { NavLink } from "react-router-dom";

class LoginCard extends Component {
  state = {
    email: "",
    password: "",
    error: ""
  };

  handleChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  onSubmit = async () => {
    let userData = { email: this.state.email, password: this.state.password };
    const url = "http://localhost:3000/api/users";
    let id;

    try {
      const user = await fetchUser(url, userData);
      id = user.data.id;
      this.props.loginUser(user.data);
    } catch (error) {
      console.log(error);
      this.setState({ error });
    }
    
    try {     
      const favorites = await fetchFavorites(`http://localhost:3000/api/users/${id}/favorites`);
      this.props.getFavorites(favorites.data)
    } catch (error) {
      this.setState({ error });
    }

    this.clearForm();
  };

  clearForm = () => {
    this.setState({
      email: "",
      password: ""
    });
  };

  render() {
    return (
      <form className="login-card">
        <input
          onChange={this.handleChange}
          name="email"
          value={this.state.email}
          type="text"
          placeholder="Email"
        />
        <input
          onChange={this.handleChange}
          name="password"
          value={this.state.password}
          type="text"
          placeholder="Password"
        />
        <NavLink to="/">
          <button type="button" onClick={this.onSubmit}>
            Login
          </button>
        </NavLink>
        <NavLink to="/register">Don't have an account? Register here.</NavLink>
      </form>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  loginUser: user => dispatch(loginUser(user)),
  getFavorites: favorites => dispatch(getFavorites(favorites))
});

export default connect(
  null,
  mapDispatchToProps
)(LoginCard);
