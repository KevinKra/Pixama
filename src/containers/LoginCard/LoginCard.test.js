import { LoginCard, mapDispatchToProps, mapStateToProps } from './LoginCard';
import { loginUser, updatePopularFavorites, updateRomanceFavorites, updateFavorites} from '../../actions';
import React from 'react';
import { shallow } from 'enzyme';
import * as apiCalls from '../../api/apiCalls';

jest.mock("../../api/apiCalls", () => ({
  fetchUser: jest.fn().mockImplementation(()=> {
    return Promise.resolve({id: 1, name: "fake", emai: "fake", password: "fake"})
  }),
  fetchFavorites: jest.fn().mockImplementation(() => {
    return Promise.resolve({title:'Toy Story'},{title: 'Fight Club'})
  })
}))

const mockLoginUser = jest.fn()

describe ('LoginCard', ()=> {
  describe('LoginCard component', () => {
  let wrapper;
  let instance;

  beforeEach(()=> {
    wrapper = shallow(<LoginCard loginUser={mockLoginUser} />)
    instance = wrapper.instance()
  })

   //UI Test
  it('should match the snapshot', () => {
    wrapper = shallow(<LoginCard />)
    expect(wrapper).toMatchSnapshot()
  })
    
   //component tests

  it('should have a default state', ()=> {
    const expected = {
      email:"",
      password:"",
      error: "",
      redirect: false
    }
    expect(wrapper.state()).toEqual(expected)
  })

  it ('should update state when handleChange is called', ()=> {
    const mockEvent = {target: {name: "email", value: "email@address.com"}}
    const expected = 'email@address.com'
    instance.handleChange(mockEvent)
    expect(wrapper.state('email')).toEqual(expected)
  })

  describe("onSubmit function", () => {

    it.skip('should call the onSubmit method when the login button is clicked', ()=>{
      //async?
  })

    it("should invoke fetchUser with the correct params", async () => {
      const url = "http://localhost:3000/api/users"
      const arg = {"email": "", "password": ""}
      await instance.onSubmit()
      expect(apiCalls.fetchUser).toHaveBeenCalledWith(url, arg)
    })

    it.skip("should invoke loginUser with the result of the fetch", async () => {
      const mockFn = jest.spyOn( instance, 'populateFavorites')
      await instance.onSubmit();
      expect(mockFn).toHaveBeenCalled(); 
    })

    it('should invoke populateFavorites with an id', ()=> {


    })

  it.skip('sets an error when the fetch fails', async () => {

    //   jest.mock("../../api/apiCalls", () => ({

    //   fetchUser: jest.fn().mockImplementation(()=> {
    //     return Promise.reject(new Error("Error fetching"))
    //   })
    // })

    const mockData = {name: "blahhhh"} 
    await  wrapper.instance().fetchUser(mockData)
    expect(wrapper.state('error')).toEqual("Email and password do not match")
    })
    
  })

  describe('clearForm', ()=> {

    it('should reset email and password in state when clearForm is called', ()=>{
      const expected = {email: "", password: "", error: "", redirect: false}
      instance.setState({email: 'pug@pugs.com', password: 'pugs'})
      instance.clearForm()
      expect(wrapper.state()).toEqual(expected)
      })
  });



describe('populateFavorites', ()=> { 
//fetch Favorite is called
//set an error
it ('should call ClearForm when onSubmit is called', async ()=> {
  instance.clearForm = jest.fn()
  await instance.populateFavorites()
  expect(instance.clearForm).toHaveBeenCalled()
})

})

  //renderRedirect

  it.skip('renderRedirect should return a Redirect is state.redirect ===true', ()=>{

  })

})

  //MDTP
  describe('mapStateToProps', ()=> { 
    
    it('should return objects with the popular, romance and favorite movies array', () => {
      const mockState = {
          currentUser: {name: "Anneke"},
          popularMovies: [{tite: "Fight Club"}],
          romanceMovies: [{tite: "Casablanca"}],
          favorites: [{tite: "Fight Club"}]
        }
        const expected = {
          popularMovies: [{tite: "Fight Club"}],
          romanceMovies: [{tite: "Casablanca"}],
          favorites: [{tite: "Fight Club"}]
        }
        const mappedProps = mapStateToProps(mockState)
  
        expect(mappedProps).toEqual(expected)
      })
      });
    
 
  

  describe('mapDispatchToProps', ()=> {
    it('calls dispatch with a loginUser action when onSubmit is called',()=>{

      const mockDispatch = jest.fn()
      const actionToDispatch = loginUser({name:"Taylor", email: "fake@test.com", password: "blahhhh"})
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.loginUser({name:"Taylor", email: "fake@test.com", password: "blahhhh"})
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('calls dispatch with a updatePopularFavorites action when onSubmit is called',()=>{
      const mockDispatch = jest.fn()
      const actionToDispatch = updatePopularFavorites([{title: "Fight Club"}])
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.updatePopularFavorites([{title: "Fight Club"}])
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('calls dispatch with a updateRomanceFavorites action when onSubmit is called',()=>{
      const mockDispatch = jest.fn()
      const actionToDispatch = updateRomanceFavorites([{title: "Fight Club"}])
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.updateRomanceFavorites([{title: "Fight Club"}])
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('calls dispatch with a updateFavorites action when onSubmit is called',()=>{
      const mockDispatch = jest.fn()
      const actionToDispatch = updateFavorites([{title: "Fight Club"}])
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.updateFavorites([{title: "Fight Club"}])
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  
  });

});


