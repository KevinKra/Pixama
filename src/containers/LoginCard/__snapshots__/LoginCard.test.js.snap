// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginCard LoginCard component should match the snapshot 1`] = `
<form
  className="login-card"
>
  <input
    name="email"
    onChange={[Function]}
    placeholder="Email"
    type="text"
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    placeholder="Password"
    type="text"
    value=""
  />
  <NavLink
    to="/"
  >
    <button
      className="submit-button"
      onClick={[Function]}
      type="button"
    >
      Login
    </button>
  </NavLink>
  <NavLink
    to="/register"
  >
    Don't have an account? Register here.
  </NavLink>
</form>
`;
